### PR TITLE
Issue/6462 pydantic v2

### PR DESCRIPTION
**Pull request by Hugo-Inmanta**
# Description

Part of https://github.com/inmanta/inmanta-core/issues/6462

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [ ] Attached issue to pull request
- [ ] Changelog entry
- [ ] Type annotations are present
- [ ] Code is clear and sufficiently documented
- [ ] No (preventable) type errors (check using make mypy or make mypy-diff)
- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)

### CODE REVIEW

**Comment by Hugo-Inmanta** [comment]
- inmanta-lsm PR: https://github.com/inmanta/inmanta-lsm/pull/1425
- lsm PR: https://github.com/inmanta/lsm/pull/640

File: src/inmanta/server/validate_filter.py [hunk 1]
```diff
@@ -121,8 +121,13 @@ def to_query_type(self) -> Optional[Tuple[QueryType, object]]:
 class BooleanEqualityFilter(Filter):
     """Represents a valid boolean which should be handled as an equality filter"""
 
-    field: Optional[bool]
-    validate_field: classmethod = validator("field", pre=True, allow_reuse=True)(parse_single_value)
+    field: Optional[bool] = None
+
+    # [pydantic v1->v2] https://docs.pydantic.dev/2.3/migration/#the-allow_reuse-keyword-argument-is-no-longer-necessary
+    @field_validator("field", mode="before")
+    @classmethod
```
Comment (Hugo-Inmanta) [inline src/inmanta/server/validate_filter.py:128]: Comment for cross-referencing

File: src/inmanta/module.py [hunk 2]
```diff
@@ -1220,13 +1221,20 @@ class Metadata(BaseModel):
     name: str
     description: Optional[str] = None
     freeze_recursive: bool = False
-    freeze_operator: str = Field(default="~=", regex=FreezeOperator.get_regex_for_validation())
+    freeze_operator: str = Field(default="~=", pattern=FreezeOperator.get_regex_for_validation())
 
-    _raw_parser: Type[RawParser]
+    # refactored _raw_parser into _raw_parser_parse because: attributes starting with underscore are converted into a
+    # "private attribute" which is not validated or even set during calls to __init__, model_validate (Source:
+    # https://docs.pydantic.dev/latest/usage/models/#private-model-attributes)
```
Comment (Hugo-Inmanta) [inline src/inmanta/module.py:1228]: Comment for cross-referencing

File: src/inmanta/data/model.py [hunk 3]
```diff
@@ -147,49 +120,49 @@ class CompileData(BaseModel):
...
-    requested: Optional[datetime.datetime]
-    started: Optional[datetime.datetime]
+    requested: Optional[datetime.datetime] = None
+    started: Optional[datetime.datetime] = None
 
     do_export: bool
     force_update: bool
     metadata: JsonType
-    environment_variables: Dict[str, str]
+    environment_variables: Dict[str, SimpleTypes]
```
Comment (bartv) [inline src/inmanta/data/model.py:131]: Is this what we want?

File: src/inmanta/agent/handler.py [hunk 4]
```diff
@@ -1019,7 +1019,7 @@ def _call_discovered_resource_create_batch(
 
             discovered_resources_raw: abc.Mapping[ResourceIdStr, TDiscovered] = self.discover_resources(ctx, resource)
             discovered_resources: abc.Sequence[DiscoveredResource] = [
-                DiscoveredResource(discovered_resource_id=resource_id, values=values)
+                DiscoveredResource(discovered_resource_id=resource_id, values=values.model_dump())
```
Comment (bartv) [inline src/inmanta/agent/handler.py:1022]: This is required to make pydantic accept the object used in the tests. However, the typing in here does not allow this. See this [thread](https://inmanta.slack.com/archives/CKRF0C8R3/p1697191108405509) for more details

File: src/inmanta/data/dataview.py [hunk 5]
```diff
@@ -678,7 +678,12 @@ def construct_dtos(self, records: Sequence[Record]) -> Sequence[model.CompileRep
                 do_export=compile["do_export"],
                 force_update=compile["force_update"],
                 metadata=json.loads(compile["metadata"]) if compile["metadata"] else {},
-                environment_variables=json.loads(compile["environment_variables"]) if compile["environment_variables"] else {},
+                environment_variables={
+                    k: str(v)
+                    for k, v in (
+                        json.loads(compile["environment_variables"]) if compile["environment_variables"] else {}
+                    ).items()
+                },
```
[the remark this answers is not in the document]
Comment (bartv) [reply src/inmanta/data/dataview.py:686]: this made a test fail. I do not remember why I did this. Make sense to revert this and see what happens.

File: tests/test_data_model.py [hunk 6]
```diff
@@ -92,8 +92,16 @@ def test_log_line_deserialization():
...
+        LogLine(level="LOUD", msg="test", args=[], kwargs={}, timestamp=datetime.datetime.now())
+    assert "is not a valid" in str(excinfo.value)
+
+    with pytest.raises(ValueError) as excinfo:
+        LogLine(level=43, msg="test", args=[], kwargs={}, timestamp=datetime.datetime.now())
+    assert "is not a valid" in str(excinfo.value)
+
+    LogLine(level=50, msg="test", args=[], kwargs={}, timestamp=datetime.datetime.now())
+
+    # TODO: This does not list what is possible? Removing the validator will
```
Comment (sanderr) [inline tests/test_data_model.py:104]: @bartv I changed it to return the following:

```
1 validation error for LogLine
level
  Value error, Input should be 'CRITICAL' | 50, 'ERROR' | 40, 'WARNING' | 30, 'INFO' | 20, 'DEBUG' | 10 or 'TRACE' | 3 [type=value_error, input_value='LOUD', input_type=str]
    For further information visit https://errors.pydantic.dev/2.4/v/value_error
```

**Pull request closed without merging**